pkg/pillar/cmd/zedrouter/patchenvelopes.go

```diff
@@ -0,0 +1,331 @@
+// means that request to update same object will be only once, so there will be no queue of
+// go routines piling up but it will take more CPU time to process it.
+// NewPatchEnvelopes() starts goroutine processStateUpdate() which reads from the channel and updates
+// currentState to desired one.
+// This way handlers can do work of determining which patch envelopes actually need change (if any)
+// and send back in go routine rest of the update including slow work.
+// Note that this channels are only accessible from the outside by calling a function which returns
+// write-only channel, meaning that updateStateNotificationCh should not be
+// read from anywhere except processStateUpdate() so that there could not be any deadlock.
+type PatchEnvelopes struct {
+	sync.Mutex
+
+	updateStateNotificationCh chan struct{}
+
+	envelopesToUpdate map[uuid.UUID]bool
+	envelopesToDelete map[uuid.UUID]bool
+
+	currentState      map[uuid.UUID]types.PatchEnvelopeInfo
+	envelopes         map[uuid.UUID]types.PatchEnvelopeInfo
+	completedVolumes  map[uuid.UUID]types.VolumeStatus
+	contentTreeStatus map[uuid.UUID]types.ContentTreeStatus
+
+	log *base.LogObject
+}
+
+// UpdateStateNotificationCh return update channel to write UUIDs of PEs
+func (pes *PatchEnvelopes) UpdateStateNotificationCh() chan<- struct{} {
+	return pes.updateStateNotificationCh
+}
+
+// NewPatchEnvelopes returns PatchEnvelopes structure and starts goroutine
+// to process messages from channels. Note that we create buffered channels
+// to avoid unbounded processing time in writing to channel
+func NewPatchEnvelopes(log *base.LogObject) *PatchEnvelopes {
+	pe := &PatchEnvelopes{
+
+		updateStateNotificationCh: make(chan struct{}, 1),
+
+		envelopesToUpdate: make(map[uuid.UUID]bool),
+		envelopesToDelete: make(map[uuid.UUID]bool),
+
+		currentState:      make(map[uuid.UUID]types.PatchEnvelopeInfo),
+		envelopes:         make(map[uuid.UUID]types.PatchEnvelopeInfo),
+		completedVolumes:  make(map[uuid.UUID]types.VolumeStatus),
+		contentTreeStatus: make(map[uuid.UUID]types.ContentTreeStatus),
+
+		log: log,
+	}
+
+	go pe.processStateUpdate()
+
+	return pe
+}
+
+func (pes *PatchEnvelopes) processStateUpdate() {
+	for {
+		select {
+		case <-pes.updateStateNotificationCh:
+			pes.updateState()
+		}
+	}
+}
+
+func (pes *PatchEnvelopes) updateState() {
+	pes.Lock()
+
+	desiredUpdates := make(map[uuid.UUID]bool, len(pes.envelopesToUpdate))
+
+	for k, v := range pes.envelopesToUpdate {
+		desiredUpdates[k] = v
+		delete(pes.envelopesToUpdate, k)
+	}
+
+	desiredDelete := make(map[uuid.UUID]bool, len(pes.envelopesToDelete))
+	for k, v := range pes.envelopesToDelete {
+		desiredDelete[k] = v
+		delete(pes.envelopesToDelete, k)
+	}
+
+	pes.Unlock()
+
+	for peUUID, toDelete := range desiredDelete {
+		if toDelete {
+			delete(pes.currentState, peUUID)
+		}
+	}
+
+	for peUUID, needsUpdate := range desiredUpdates {
+		if needsUpdate {
+			peInfo := pes.envelopes[peUUID]
```

Comment:
Is envelopes read by some other goroutine? If so you need to serialize all reads and writes to that map.

Reply:
Yes, we call it from updateState() and from UpdateVolumeStatus() and UpdateContentTreeStatus(), because we are determining which PE are affected. So in theory several goroutines can try to access it, that's why I wrapped it into LockedMap